clang/lib/Sema/SemaDecl.cpp

```diff
@@ -4754,6 +4754,11 @@ void Sema::MergeVarDecl(VarDecl *New, LookupResult &Previous) {
     return New->setInvalidDecl();
   }
 
+  if (Old->getFormalLinkage() != New->getFormalLinkage()) {
+    Diag(New->getLocation(), diag::err_multiple_linkage) << New->getDeclName();
+    return New->setInvalidDecl();
+  }
```

Comment:
I don't think this logic is correct. See C23 6.2.2 (especially p4): 

> For an identifier declared with the storage-class specifier extern in a scope in which a prior declaration of that identifier is visible, if the prior declaration specifies internal or external linkage, the linkage of the identifier at the later declaration is the same as the linkage specified at the prior declaration. 

So a mismatch between linkages isn't always an issue to be diagnosed: https://godbolt.org/z/rx7rGhv7c (notice how the local variable `x` is not an issue despite having different linkage from the `extern` declaration: https://godbolt.org/z/Yjqo4eY9M) This is possibly why you're getting as many CI failures as you're seeing.

Reply:
@AaronBallman
> If the prior declaration specifies internal or external linkage, the linkage of the identifier at the later declaration is the same as the linkage specified at the prior declaration.

Then when should the declaration be an issue (https://godbolt.org/z/roE9xh76d)?
[Note: the diagnostic of clang and GCC is different.]
If the external linkage follows the previous linkage, when could the error happen?

Reply:
I think http://eel.is/c++draft/dcl.stc#example-1 has a lot of good examples that are worth looking at.

We do a lot of this checking already, starting from https://github.com/llvm/llvm-project/blob/cb8d83a77c25e529f58eba17bb1ec76069a04e90/clang/lib/Sema/SemaDecl.cpp#L4643 and the issue might relate to the FIXME at https://github.com/llvm/llvm-project/blob/cb8d83a77c25e529f58eba17bb1ec76069a04e90/clang/lib/Sema/SemaDecl.cpp#L4696